File: src/helm-tools.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js"
import { keyValueSetArgs, kubectlContextArgs, namespaceArgs } from "./args.js"
import { HelmListInput, HelmReleaseInput, HelmTemplateInput } from "./schemas.js"
import { commandTool } from "./tool-runner.js"
import type { RuntimeConfig } from "./types.js"

export function registerHelmTools(server: McpServer, config: RuntimeConfig): void {
  server.registerTool(
    "helm_list",
    { description: "List Helm releases for an explicit EKS cluster context.", inputSchema: HelmListInput.shape },
    ({ cluster, namespace, allNamespaces }) =>
      commandTool(config, {
        command: "helm",
        args: ["list", "--kube-context", cluster, ...(allNamespaces ? ["--all-namespaces"] : namespaceArgs(namespace)), "--output", "json"]
      })
  )

  server.registerTool(
    "helm_status",
    { description: "Read Helm release status.", inputSchema: HelmReleaseInput.shape },
    ({ cluster, namespace, release }) =>
      commandTool(config, {
        command: "helm",
        args: ["status", release, "--kube-context", cluster, ...namespaceArgs(namespace), "--output", "json"]
      })
  )

  server.registerTool(
    "helm_history",
    { description: "Read Helm release history.", inputSchema: HelmReleaseInput.shape },
    ({ cluster, namespace, release }) =>
      commandTool(config, {
        command: "helm",
        args: ["history", release, "--kube-context", cluster, ...namespaceArgs(namespace), "--output", "json"]
      })
  )

  server.registerTool(
    "helm_template",
    { description: "Render a Helm chart locally without installing it.", inputSchema: HelmTemplateInput.shape },
    ({ cluster, namespace, release, chart, valuesFile, set }) =>
      commandTool(config, {
        command: "helm",
        args: [
          "template",
          release,
          chart,
          "--kube-context",
          ...kubectlContextArgs(cluster).slice(1),
          ...namespaceArgs(namespace),
          ...(valuesFile === undefined ? [] : ["--values", valuesFile]),
          ...keyValueSetArgs(set)
        ]
      })
  )

  server.registerTool(
    "helm_lint",
    { description: "Run helm lint locally against a chart.", inputSchema: HelmTemplateInput.shape },
    ({ namespace, chart, valuesFile, set }) =>
      commandTool(config, {
        command: "helm",
        args: ["lint", chart, ...namespaceArgs(namespace), ...(valuesFile === undefined ? [] : ["--values", valuesFile]), ...keyValueSetArgs(set)]
      })
  )
}

File: safe_eks_mcp/helm_tools.py
from __future__ import annotations

from .args import key_value_set_args, namespace_args
from .runner import command_tool
from .types import CliInvocation, RuntimeConfig, ToolPayloadJson
from .validation import (
    normalize_set_values,
    optional_namespace,
    require_cluster,
    require_name,
    require_path,
    require_region,
)


def helm_list(
    config: RuntimeConfig, region: str, cluster: str, namespace: str | None = None, allNamespaces: bool = False
) -> ToolPayloadJson:
    _ = require_region(region)
    args = ["list", "--kube-context", require_cluster(cluster)]
    args.extend(("--all-namespaces",) if allNamespaces else namespace_args(optional_namespace(namespace)))
    args.extend(("--output", "json"))
    return command_tool(config, CliInvocation(command="helm", args=tuple(args)))


def helm_status(
    config: RuntimeConfig, region: str, cluster: str, release: str, namespace: str | None = None
) -> ToolPayloadJson:
    _ = require_region(region)
    return command_tool(
        config,
        CliInvocation(
            command="helm",
            args=(
                "status",
                require_name(release, "release"),
                "--kube-context",
                require_cluster(cluster),
                *namespace_args(optional_namespace(namespace)),
                "--output",
                "json",
            ),
        ),
    )


def helm_history(
    config: RuntimeConfig, region: str, cluster: str, release: str, namespace: str | None = None
) -> ToolPayloadJson:
    _ = require_region(region)
    return command_tool(
        config,
        CliInvocation(
            command="helm",
            args=(
                "history",
                require_name(release, "release"),
                "--kube-context",
                require_cluster(cluster),
                *namespace_args(optional_namespace(namespace)),
                "--output",
                "json",
            ),
        ),
    )


def helm_template(
    config: RuntimeConfig,
    region: str,
    cluster: str,
    release: str,
    chart: str,
    namespace: str | None = None,
    valuesFile: str | None = None,
    set: dict[str, str | int | bool] | None = None,
) -> ToolPayloadJson:
    _ = require_region(region)
    args = [
        "template",
        require_name(release, "release"),
        require_path(chart, "chart"),
        "--kube-context",
        require_cluster(cluster),
        *namespace_args(optional_namespace(namespace)),
    ]
    if valuesFile is not None:
        args.extend(("--values", require_path(valuesFile, "valuesFile")))
    args.extend(key_value_set_args(normalize_set_values(set)))
    return command_tool(config, CliInvocation(command="helm", args=tuple(args)))


def helm_lint(
    config: RuntimeConfig,
    region: str,
    cluster: str,
    release: str,
    chart: str,
    namespace: str | None = None,
    valuesFile: str | None = None,
    set: dict[str, str | int | bool] | None = None,
) -> ToolPayloadJson:
    _ = require_region(region)
    _ = require_cluster(cluster)
    _ = require_name(release, "release")
    args = ["lint", require_path(chart, "chart"), *namespace_args(optional_namespace(namespace))]
    if valuesFile is not None:
        args.extend(("--values", require_path(valuesFile, "valuesFile")))
    args.extend(key_value_set_args(normalize_set_values(set)))
    return command_tool(config, CliInvocation(command="helm", args=tuple(args)))
